library-best-practices/JLBP-3.md

```diff
@@ -0,0 +1,47 @@
+[JLBP-3] Use Semantic versioning
+--------------------------------
+
+Definition of semver: https://semver.org
+
+- For the purpose of compliance with semver, the stable surface of a library
+  serves as its "Public API." Annotations can be used to mark features
+  (classes, methods, etc) as unstable, and thus not part of the Public API,
+  such that semver rules don't have to apply. 
+  - Examples of annotations:
+    - Guava uses `@Beta`
+    - grpc-java uses `@ExperimentalApi`
+    - Google-cloud-java uses `@BetaApi`, `@InternalApi`, and
+      `@InternalExtensionOnly`
+  - Library documentation should point users to a tool (or tools) that can
+    help them automatically detect when they are inadvertently using features
+    marked with these annotations.
+  - Even though semver rules don't apply to unstable features, it is
+    recommended to bump a library's minor version if unstable features have
+    surface breakages.
+- Tools are available to help identify accidental incompatibilities within a
+  major version. Examples:
+  - https://lvc.github.io/japi-compliance-checker/ 
+  - http://www.mojohaus.org/clirr-maven-plugin/
+  - Example compatibility report for grpc-core:
+    https://abi-laboratory.pro/index.php?view=timeline&lang=java&l=grpc-core 
+- Examples of breaking changes to a stable Public API that require a new
+  major version:
+  - Upgrading to a non-compatible dependency that is exposed through a
+    library's Public API (for dependencies that follow semver, this happens
+    when a dependency is bumped to a higher major version)
+  - Changing a method signature
+  - Removing a deprecated method
```

Comment:
Removing a method (that is, deprecated or not)

Reply:
Done